File: tsmodules/src/discord.ts
import { errInternal } from './errors';
import { DiscordAccessToken, DiscordUser } from './types';

export function discordExchangeCode(ctx: nkruntime.Context, nk: nkruntime.Nakama, logger: nkruntime.Logger, oauthCode: string, oauthRedirectUrl: string): DiscordAccessToken {
  var params = `client_id=${ctx.env["DISCORD_CLIENT_ID"]}&` +
    `client_secret=${ctx.env["DISCORD_CLIENT_SECRET"]}&` +
    `code=${oauthCode}&` +
    `grant_type=authorization_code&` +
    `redirect_uri=${oauthRedirectUrl}&` +
    `scope=identify`;
  
  // exchange the oauthCode for the user's access token
  let response = nk.httpRequest("https://discord.com/api/v10/oauth2/token", "post",
    {
      'Accept': 'application/json',
      "Content-Type": "application/x-www-form-urlencoded",
      "Authorization": `${ctx.env["DISCORD_CLIENT_ID"]}:${ctx.env["DISCORD_CLIENT_SECRET"]}`,
    }, params);

  if (response.code != 200) {
    if ("error" in response) {
      logger.error("Discord code exchange failed: %s", response.body);
      throw {
        message: `Discord code exchange failed: ${response.body}`,
        code: nkruntime.Codes.UNAUTHENTICATED
      } as nkruntime.Error;
    }
  }

  try {
    return JSON.parse(response.body) as DiscordAccessToken;

  } catch (error) {
    logger.error("Could not decode discord response body: %s", response.body);
    throw errInternal(`Could not decode discord response body: ${response.body}`);
  }
}


export function discordRefreshAccessToken(ctx: nkruntime.Context, nk: nkruntime.Nakama, logger: nkruntime.Logger, accessToken: DiscordAccessToken): DiscordAccessToken {
  var params = `grant_type=refresh_token&refresh_token=${accessToken.refresh_token}`;


  let response = nk.httpRequest("https://discord.com/api/v10/oauth2/token", "post",
    {
      'Accept': 'application/json',
      "Content-Type": "application/x-www-form-urlencoded",
      "Authorization": `${ctx.env["DISCORD_CLIENT_ID"]}:${ctx.env["DISCORD_CLIENT_SECRET"]}`,
    }, params);

  let responseJson = null;
  try {
    responseJson = JSON.parse(response.body);
  } catch (error) {
    logger.error("Could not decode discord response body: %s", response.body);
    throw errInternal(`Could not decode discord response body: ${response.body}`);
  }

  if (response.code != 200) {
    if ("error" in responseJson) {
      logger.error("Discord code exchange failed: %s", response.body);
      throw {
        message: `Discord code exchange failed: ${response.body}`,
        code: nkruntime.Codes.UNAUTHENTICATED
      } as nkruntime.Error;
    }
  }
  return responseJson as DiscordAccessToken;
}


export function discordGetCurrentUser(ctx: nkruntime.Context, nk: nkruntime.Nakama, logger: nkruntime.Logger, accessToken: DiscordAccessToken): DiscordUser {

  let response = nk.httpRequest("https://discord.com/api/v10/users/@me", "get",
    {
      "Content-Type": "application/x-www-form-urlencoded",
      "Authorization": `Bearer ${accessToken.access_token}`,
      'Accept': 'application/json'
    }, null);

  
  if (response.code != 200) {
    logger.error("Discord user lookup failed: %s", response.body);
    throw errInternal(`Discord user lookup failed: ${response.body}`);
  }

  return JSON.parse(response.body) as DiscordUser;
}

File: gomodules/server/services/login/discord.go
package login

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

type DiscordSecrets struct {
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
}

// DiscordAccessToken represents the Discord access token structure.
type DiscordAccessToken struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"`
	Scope        string `json:"scope"`
	Secrets      DiscordSecrets
}

func (t *DiscordAccessToken) Marshal() ([]byte, error) {
	return json.Marshal(t)
}

func UnmarshalDiscordAccessToken(data []byte, s *DiscordSecrets) (DiscordAccessToken, error) {
	r := DiscordAccessToken{
		Secrets: *s,
	}

	err := json.Unmarshal(data, &r)
	return r, err
}

func (t *DiscordAccessToken) Refresh() (*DiscordAccessToken, error) {
	url := "https://discord.com/api/v10/oauth2/token"
	params := fmt.Sprintf("grant_type=refresh_token&refresh_token=%s", t.RefreshToken)

	req, err := http.NewRequest("POST", url, bytes.NewBufferString(params))
	if err != nil {
		return t, err
	}

	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", fmt.Sprintf("Basic %s:%s", t.Secrets.ClientID, t.Secrets.ClientSecret))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return t, err
	}
	if resp.StatusCode != http.StatusOK {
		return t, fmt.Errorf("discord refresh failed: %s", resp.Status)
	}
	defer resp.Body.Close()

	if err := json.NewDecoder(resp.Body).Decode(&t); err != nil {
		return t, err
	}

	return t, nil
}

// DiscordUser represents the Discord user structure.
type DiscordUser struct {
	ID                 string `json:"id"`
	Username           string `json:"username"`
	Discriminator      string `json:"discriminator"`
	Avatar             string `json:"avatar"`
	DiscordAccessToken *DiscordAccessToken
}

// discordGetCurrentUser gets the current Discord user.
func (u *DiscordUser) User() (*DiscordUser, error) {
	url := "https://discord.com/api/v10/users/@me"

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return u, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", "Bearer "+u.DiscordAccessToken.AccessToken)
	req.Header.Set("Accept", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return u, err
	}
	defer resp.Body.Close()

	if err := json.NewDecoder(resp.Body).Decode(&u); err != nil {
		return u, err
	}

	if resp.StatusCode != http.StatusOK {
		return u, fmt.Errorf("discord user lookup failed: %s", resp.Status)
	}

	return u, nil
}
